proposals/span-safety.md

```diff
@@ -0,0 +1,104 @@
+
+An additional problem arises due to composition. It would be generally desirable to build more complex data types that would embed `Span<T>` instances. Such composite types would have to be structs and would share all the hazards and requirements of `Span<T>`. As a result the safety rules described here should be viewed as applicable to the whole range of **_ref-like types_**
+
+At the minimum the restrictions described below would need to apply to `Span<T>`.
+
+## `ref-like` types must be stack-only. ##
+
+C# compiler already has a concept of a restricted types that covers special platform types such as `TypedReference`. In some cases those types are referred as ref-like types as well. We should probably use some different term to refer to `TypedReference` and similar types - like `restricted types` to avoid confusion.
+
+In this document the "ref like" types include only `Span<T>` and related types. And the ref-like variables mean variables of "ref-like" types.  
+
+In order to force ref-like variables to be stack only, we need the following restrictions: 
+
+- ref-like type cannot be a type of an array element
+- ref-like type cannot be used as a generic type argument
+- ref-like variable cannot be boxed
+- ref-like type cannot be a field of ordinary not ref-like type
+- indirect restrictions, such as disallowed use of ref-like types in async methods, which are really a result of disallowing ref-like typed fields.
+
+Note:
+
+- it is ok to allow returning ref-like values from methods (additional rules apply when referring to local data)
+- it is ok for a ref-like type be a field of another struct as long as the struct itself is ref-like.
+
+
+## ref-like variables can be passed as `in` parameters ##
+
+- it is ok to pass a ref-like variable by reference to a call as long as the reference is **_readonly_** . 
+The reason is that exposing ref-like variables via writeable references make it nearly impossible to reason about their safety.
+
+Example:
+
+```cs
+
+Span<T> Caller()
+{
+	Span<T> safeToReturn = GetSafeToReturnSpan(array);
+
+	if (TodaIsNotFriday())
+	{
+		// callee can "poison" any assumption that we have made about safeToReturn
+		Callee(ref safeToReturn);
+	}
+
+	// is this safe?
+	return safeToReturn;
+}
+
+void Callee(ref Span<int> arg1)
+{
+	// local references are escaping. 
+	// what rule would disallow this?
+	arg1 = new Span<T>(localloc(...));
+}
+
+```
+
+Note: returning by a ordinary writeable reference appears to be ok by itself, but with values on heap being impossible, values in local frame being not returnable and variables passed from the caller being readonly, it may not matter.
+
+Another way to look at this is that writeable references to a ref-like variable would never be exposed outside of the containing method, except, of course, if that method is a constructor which has to take `this` by a writeable ref.
+
+## `ref-like` types must be readonly structs. ##
+
+- ref-like types must be readonly. 
```

Comment:
@VSadov - `stackalloc` requires `Foo` to be `unsafe`, which probably means that `Foo` can do unsafe things with `this` even if the type is readonly. This might be a better example:

```csharp
struct MyStruct
{
    readonly ref int _x;

    public MyStruct(ref int x)
    {
        ref _x = ref x;
    }

    public void Foo()
    {
        int y = 123;
        this = new MyStruct(ref y);
    }

    public MyStruct Bar()
    {
        int y = 123;
        return new MyStruct(ref y);
    }
}
```

To me it seems that assigning to `this` should follow the same rules as returning a `ref`, i.e., the assignment in `Bar` should be rejected for the same reason as the return statement in `Baz`.